### PR TITLE
159 switch emails to cdsid for users

### PR DESCRIPTION
**Pull request by jwcherney**
## Issue
Resolves 159

## What was done
- [x] Add API Endpoint to add users as editors to a space
- [x] Update UI to accept CDSIDs as input for Editors
- [x] Add/use feature toggle (#perm) for updated UI
- [x] Refactor and reorganize to make switchover to new UI (without feature toggle) easier

## How to test
1. Deploy to test environment
2. Create/open a space
3. Select Share Access
4. Enter one or more CDSIDs 
5. Click Invite
Alternate paths: 
- CDSIDs can be copy/pasted in
- Invalid entries (that don't conform to CDSID format) are kept as strings, an error message is shown, and the Invite button is disabled. 
- Text entry area will expand as many CDSIDs are added

### Testing
- [ ] Add jest-axe unit tests (for new ui components)

### Manual Checks
- [ ] Text and background color meets a minimum color contrast ratio of 4.5:1
- [ ] All non-text content (e.g. images, icon) has alt text
- [ ] Web page includes appropriate headings to communicate structure/hierarchy (h1>h2>h3)
- [ ] All functionality is available to a keyboard
  - [ ] Focus styles are highly visible 
  - [ ] Tab order is logical and aligns with the visual order 
- [ ] Voice over is understandable and logical
  - [ ] Aria labels are added to elements with no visible text 
  - [ ] Decorative elements (e.g. non-informative icons) are hidden from screen readers  
- [ ] Form fields have appropriately coded labels that are visible during input 
- [ ] Links are visually identifiable and have a clear focus and active state

### CODE REVIEW

**Comment by jwcherney** [comment]
This is a large number of changes. It would probably be best if we do a code walk-through either prior to or along with the code review.

File: api/src/test/kotlin/com/ford/internalprojects/peoplemover/user/UserServiceTest.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
+package com.ford.internalprojects.peoplemover.user
```
Comment (dulcimerist) [inline api/src/test/kotlin/com/ford/internalprojects/peoplemover/user/UserServiceTest.kt:1]: No copyright header

File: api/src/test/kotlin/com/ford/internalprojects/peoplemover/user/UserControllerApiTest.kt [hunk 2]
```diff
@@ -0,0 +1,423 @@
+package com.ford.internalprojects.peoplemover.user
```
Comment (dulcimerist) [inline api/src/test/kotlin/com/ford/internalprojects/peoplemover/user/UserControllerApiTest.kt:1]: No copyright header

File: api/src/main/kotlin/com/ford/internalprojects/peoplemover/utilities/ListPattern.kt [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package com.ford.internalprojects.peoplemover.utilities
```
Comment (dulcimerist) [inline api/src/main/kotlin/com/ford/internalprojects/peoplemover/utilities/ListPattern.kt:1]: No copyright header

File: api/src/main/kotlin/com/ford/internalprojects/peoplemover/user/UserService.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
+package com.ford.internalprojects.peoplemover.user
```
Comment (dulcimerist) [inline api/src/main/kotlin/com/ford/internalprojects/peoplemover/user/UserService.kt:1]: No copyright header

File: api/src/main/kotlin/com/ford/internalprojects/peoplemover/user/UserController.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
+package com.ford.internalprojects.peoplemover.user
```
Comment (dulcimerist) [inline api/src/main/kotlin/com/ford/internalprojects/peoplemover/user/UserController.kt:1]: No copyright header

File: ui/src/AccountDropdown/InviteEditorsFormSection.tsx [hunk 6]
```diff
@@ -43,29 +69,84 @@ const getUsers = (currentSpace: Space, setUsersList: (usersList: UserSpaceMappin
...
+                .finally(() => {
+                    setCurrentModal({modal: AvailableModals.GRANT_EDIT_ACCESS_CONFIRMATION});
+                });
+        }
+    };
+
+    useEffect(() => {
+        const enable = (invitedUserIds.length > 0 && inputValue.trim().length === 0)
+                || (!!inputValue.trim().match(userIdPattern))
+                || (invitedUserEmails.length > 0);
```
Comment (dulcimerist) [inline ui/src/AccountDropdown/InviteEditorsFormSection.tsx:111]: Woof. This is an ugly conditional

File: ui/src/Utils/userIdValidator.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    options: [],
+    notValid: '',
+});
+
+export function makeOption(name: string): Option {
+    return {value: name, label: name};
+}
+
+export const nameSplitPattern = new RegExp(/,|;|\s/);
+export const userIdPattern = new RegExp(/^[a-zA-Z][a-zA-Z0-9]{1,8}$/);
```
Comment (dulcimerist) [inline ui/src/Utils/userIdValidator.ts:18]: I've seen this regex in a couple of places now - can we consolidate into one place?

File: ui/src/Utils/userIdValidator.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
+import {Option} from '../CommonTypes/Option';
```
Comment (dulcimerist) [inline ui/src/Utils/userIdValidator.ts:1]: No copyright header

File: ui/src/Utils/userIdValidator.test.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
+import {emptyValidateUserResult, makeOption, validate} from './userIdValidator';
```
Comment (dulcimerist) [inline ui/src/Utils/userIdValidator.test.ts:1]: No copyright header

File: ui/src/AccountDropdown/InviteEditorsFormSection.tsx [hunk 10]
```diff
@@ -89,22 +170,50 @@ function InviteEditorsFormSection({collapsed, currentSpace, currentUser, closeMo
...
+                                hideSelectedOptions={true}
+                                isClearable={false}
+                                components={components}
+                                hidden={collapsed}
+                                value={invitedUserIds}
+                                options={invitedUserIds}
+                                onChange={onChange}
+                                onInputChange={onInputChange}
+                                inputValue={inputValue}
+                                onKeyDown={handleKeyDownEvent}
```
Comment (dulcimerist) [inline ui/src/AccountDropdown/InviteEditorsFormSection.tsx:197]: We may want to make this onKeyUp instead

File: ui/src/Utils/userIdValidator.test.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+            ], notValid: ''});
+    });
+    it('5should return a list of valid cdsid when valid multiple passed', function() {
+        expect(validate('cdsid;cdsid1; cdsid2')).toEqual({
+            options: [makeOption('cdsid'),
+                makeOption('cdsid1'),
+                makeOption('cdsid2'),
+            ], notValid: ''});
+    });
+    it('6should return a list of valid cdsid when valid multiple passed', function() {
```
Comment (dulcimerist) [inline ui/src/Utils/userIdValidator.test.ts:47]: Could we better describe what each of these 6 is testing?

**Review by dulcimerist — CHANGES_REQUESTED** [review]
Functionality works, generally. LGTM other than my comments

**Comment by jwcherney** [comment]
To do: Add card to move cdsid (user id) validation out to an environment variable to make code more open source friendly, for other organizations that don't use cdsids.

File: ui/src/AccountDropdown/InviteEditorsFormSection.tsx (hunk 10 above)
Comment (jwcherney) [reply ui/src/AccountDropdown/InviteEditorsFormSection.tsx:197]: Test this: Change onKeyDown to onSubmit, which handles key presses (particularly Enter) automatically.

File: ui/src/Utils/userIdValidator.ts (hunk 7 above)
Comment (jwcherney) [reply ui/src/Utils/userIdValidator.ts:18]: To do: Add card to move cdsid (user id) validation out to an environment variable to make code more open source friendly, for other organizations that don't use cdsids.

File: ui/src/AccountDropdown/InviteEditorsFormSection.tsx (hunk 6 above)
Comment (jwcherney) [reply ui/src/AccountDropdown/InviteEditorsFormSection.tsx:111]: To do: Add TODO comments in code to identify code that can be removed when we make change this code from feature toggled to live. (Removing the email entries)

**Review by jodee — APPROVED** [review]
Design looks good! Works as expected.

**Review by jodee — APPROVED** [review]

**Review by dulcimerist — APPROVED** [review]
LGTM

**Pull request merged**